plugins/aws/provisioner.go

```diff
@@ -2,42 +2,155 @@ package aws
-	return p.envVarProvisioner.Description()
+	if p.selectedProvisioner != nil {
+		return p.selectedProvisioner.Description()
+	}
+	return "Provisions AWS with temporary credentials."
+}
+
+func (p *awsProvisioner) selectCorrespondingProvisioner(itemFields map[sdk.FieldName]string, out *sdk.ProvisionOutput) sdk.Provisioner {
+	awsConfig := getAWSConfiguration(out)
+	if len(out.Diagnostics.Errors) > 0 {
+		return nil
+	}
+
+	err := errorIfUnsupportedConfigurationIsPresent(awsConfig)
+	if err != nil {
+		out.AddError(err)
+		return nil
+	}
+
+	selectedProvisioner := selectProvisionerBasedOnPresentValues(itemFields, awsConfig, out)
+
+	p.selectedProvisioner = selectedProvisioner
+
+	return selectedProvisioner
+}
+
+func getAWSConfiguration(out *sdk.ProvisionOutput) *confighelpers.Config {
+	for i, arg := range out.CommandLine {
+		if arg == "--profile" {
+			if i+1 == len(out.CommandLine) {
+				out.AddError(fmt.Errorf("--profile flag was specified without a value"))
+				return nil
+			}
+			config := getAWSAuthConfigurationForProfile(out.CommandLine[i+1], out)
+			// Remove --profile flag so aws cli doesn't attempt to assume role by itself
+			out.CommandLine = removeProfileFlagFromArgs(i, out.CommandLine)
+			return config
+		}
+	}
+
+	if val := os.Getenv("AWS_PROFILE"); val != "" {
+		return getAWSAuthConfigurationForProfile(val, out)
+	}
+
+	return getAWSAuthConfigurationForProfile("default", out)
+}
+
+func getAWSAuthConfigurationForProfile(profile string, out *sdk.ProvisionOutput) *confighelpers.Config {
+	// Read config file from the location set in AWS_CONFIG_FILE env var or from  ~/.aws/config
+	configFile, err := confighelpers.LoadConfigFromEnv()
+	if err != nil {
+		out.AddError(err)
+		return nil
+	}
+
+	configLoader := confighelpers.ConfigLoader{
+		File:          configFile,
+		ActiveProfile: profile,
+	}
+
+	configuration, err := configLoader.LoadFromProfile(profile)
+	if err != nil {
+		out.AddError(err)
+		return nil
+	}
+
+	return configuration
+}
+
+func selectProvisionerBasedOnPresentValues(itemFields map[sdk.FieldName]string, awsConfig *confighelpers.Config, out *sdk.ProvisionOutput) sdk.Provisioner {
+	totp, hasTotp := itemFields[fieldname.OneTimePassword]
+	mfaSerial, hasMFASerial := itemFields[fieldname.MFASerial]
+
+	// Give priority to the mfa serial specified in 1Password
+	if hasMFASerial && awsConfig.HasMfaSerial() && awsConfig.MfaSerial != mfaSerial {
+		out.AddError(fmt.Errorf("your local AWS configuration has a different MFA serial than the one specified in 1Password"))
+	}
+	if !hasMFASerial {
+		mfaSerial = awsConfig.MfaSerial
+	}
+
+	if awsConfig != nil && awsConfig.RoleARN != "" {
+		return newAssumeRoleProvisioner(mfaSerial, totp, awsConfig)
+	}
+
+	if hasTotp && mfaSerial != "" {
+		return newMFAProvisioner(mfaSerial, totp, awsConfig.Region, awsConfig.NonChainedGetSessionTokenDuration)
+	}
+
+	return provision.EnvVarProvisioner{
+		Schema: defaultEnvVarMapping,
+	}
+}
```

Comment:
I don't really see the value of the split between the MFA provisioner and role provisioner. It probably makes more sense to merge those two into a single `STSProvisioner` which takes an optional `profile` arg, and then have an `AWSCLIProvisioner` as a lightweight wrapper that captures the `--profile` flag and passes it down to the `STSProvisioner`. What do you think?

Reply:
I like this suggestion. I think it would clean up this implementation

Reply:
Refactored accordingly, thanks!